rclcpp_action/src/server_goal_handle.cpp

```diff
@@ -12,4 +12,140 @@
 // limitations under the License.
 
+#include <rcl_action/action_server.h>
+#include <rcl_action/goal_handle.h>
+
 #include <rclcpp_action/server_goal_handle.hpp>
+#include <rclcpp/exceptions.hpp>
+
+#include <memory>
+
+namespace rclcpp_action
+{
+ServerGoalHandleBase::~ServerGoalHandleBase()
+{
+}
+
+bool
+ServerGoalHandleBase::is_canceling() const
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  rcl_action_goal_state_t state = GOAL_STATE_UNKNOWN;
+  rcl_ret_t ret = rcl_action_goal_handle_get_status(rcl_handle_.get(), &state);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "Failed to get goal handle state");
+  }
+  return GOAL_STATE_CANCELING == state;
+}
+
+bool
+ServerGoalHandleBase::is_active() const
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  return rcl_action_goal_handle_is_active(rcl_handle_.get());
+}
+
+bool
+ServerGoalHandleBase::is_executing() const
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  rcl_action_goal_state_t state = GOAL_STATE_UNKNOWN;
+  rcl_ret_t ret = rcl_action_goal_handle_get_status(rcl_handle_.get(), &state);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "Failed to get goal handle state");
+  }
+  return GOAL_STATE_EXECUTING == state;
+}
+
+void
+ServerGoalHandleBase::_set_aborted()
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  rcl_ret_t ret = rcl_action_update_goal_state(rcl_handle_.get(), GOAL_EVENT_SET_ABORTED);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+void
+ServerGoalHandleBase::_set_succeeded()
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  rcl_ret_t ret = rcl_action_update_goal_state(rcl_handle_.get(), GOAL_EVENT_SET_SUCCEEDED);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+void
+ServerGoalHandleBase::_set_canceling()
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  rcl_ret_t ret = rcl_action_update_goal_state(rcl_handle_.get(), GOAL_EVENT_CANCEL);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+void
+ServerGoalHandleBase::_set_canceled()
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  rcl_ret_t ret = rcl_action_update_goal_state(rcl_handle_.get(), GOAL_EVENT_SET_CANCELED);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+void
+ServerGoalHandleBase::_set_executing()
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  rcl_ret_t ret = rcl_action_update_goal_state(rcl_handle_.get(), GOAL_EVENT_EXECUTE);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+bool
+ServerGoalHandleBase::try_canceling() noexcept
+{
+  std::lock_guard<std::mutex> lock(rcl_handle_mutex_);
+  // Check if the goal reached a terminal state already
+  const bool active = rcl_action_goal_handle_is_active(rcl_handle_.get());
```

Comment:
We can use a short-cut and check if the goal is cancelable:

```c
const bool is_cancelable = rcl_action_goal_handle_is_cancelable(rcl_handle.get());
```

Reply:
What's the short cut? IIUC it looks like the code below would be the same.

Reply:
I think this is equivalent to the current implementation:

```c
const bool is_cancelable = rcl_action_goal_handle_is_cancelable(rcl_handle_.get());
if (is_cancelable) {
  ret = rcl_action_update_goal_state(rcl_handle_.get(), GOAL_EVENT_CANCEL);
  if (RCL_RET_OK != ret) {
    return false;
  }
}

// Get the state again
ret = rcl_action_goal_handle_get_status(rcl_handle_.get(), &state);
if (RCL_RET_OK != ret) {
  return false;
}

// If it's canceling, cancel it
if (GOAL_STATE_CANCELING == state) {
  ret = rcl_action_update_goal_state(rcl_handle_.get(), GOAL_EVENT_SET_CANCELED);
  return RCL_RET_OK == ret;
}
return false;
```